Add code, documentation and publishing information

vsbogd opened the pull request:
It is a [part of the hyperon-experimental repo](https://github.com/trueagi-io/hyperon-experimental/blob/main/lib/src/common/holeyvec.rs) which can be published as a separated data structure.

vsbogd committed:
Add vector with holes implementation with docs and test examples

vsbogd committed:
Add README.md and information to publish on crates.io

vsbogd commented:
@luketpeterson please take a look

luketpeterson commented:
I will go over it thoroughly soon. (in the next 2 hours)

One thing I like to do, although you don't have to do it if you don't want to, is that I usually use the ReadMe.md as the front page of the documentation.  You can put:
```
#![doc = include_str!("../README.md")]
```
at the top of lib.rs to accomplish this.  Although it means your README file needs to be formatted so that it is acceptable to rustdoc's parser and the generic markdown parsers.

luketpeterson commented on README.md:
```diff
@@ -0,0 +1,26 @@
+/// # Vector with holes
```
These /// are causing formatting errors in all the MD parsers I've tried.  They're not needed in the ReadMe, even when the ReadMe is incorporated into the rustdoc

vsbogd replied:
Yeah, sorry, somehow I committed part of them removed and part of them not.

luketpeterson commented on Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+description = "Vector with holes implementation. Allows removing vector's elements in a constant time."
+repository = "https://github.com/vsbogd/holeyvec"
+readme = "README.md"
+categories = [ "data-structures" ]
```
You can use up to 5 key words and 5 category slugs.  I usually use more because it increases the changes the crate will be found.

Key words can be anything.  Personally I might add "constant-time", "stable-index", and "Vec", but you may have better ideas.

Acceptable Categories are here:  https://crates.io/category_slugs

I might add "algorithms" and "rust-patterns", but it's up to you.

vsbogd replied:
Added your suggestions in https://github.com/vsbogd/holeyvec/pull/1/commits/43b82d5a129c965207a5cbf471093fe511bdd27a

luketpeterson commented on Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+name = "holeyvec"
+version = "0.1.0"
+edition = "2021"
+license-file = "LICENSE"
```
You can also specify the license directly.  I don't know if that's better than specifying a license file.  But given it's a standard license it might help crates.io.

```
license = "MIT"
```

vsbogd replied:
I used separate file because SingularityNET Foundation is mentioned in it.

luketpeterson commented on README.md:
```diff
@@ -0,0 +1,26 @@
+/// 
+/// Vector with holes data structure. Each time an element is removed its index is added into a
+/// list of empty elements. No other values are moved or touched. This allows removing elements in
+/// a constant time.
```
I might also say something like:
"And allows the indices of the other elements to remain stable."

vsbogd replied:
Thanks Luke. Fixed in https://github.com/vsbogd/holeyvec/pull/1/commits/70d003ed6ba15a1f31bbebe48beca0f83080dd67

luketpeterson commented on src/holeyvec.rs:
```diff
@@ -0,0 +1,366 @@
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter_mut()
+    }
+}
```
Here is another convenience implementation so you can collect into a HoleyVec:
```rust
impl<T> FromIterator<T> for HoleyVec<T> {
    fn from_iter<I: IntoIterator<Item=T>>(iter: I) -> Self {
        let vec: Vec<Cell<T>> = iter.into_iter().map(|t| Cell::Value(t)).collect();
        Self {first_hole: vec.len(), vec }
    }
}
```

vsbogd replied:
Thank you! Added in https://github.com/vsbogd/holeyvec/pull/1/commits/b2771f367692e92fca631195c3d3ffaa625365f5

luketpeterson commented on README.md:
````diff
@@ -0,0 +1,26 @@
+/// 
+/// ## Examples
+///
+/// ```rs
````
```suggestion
```rust
```
More markdown parsers know `rust` than `rs` as the language formatting identifier

vsbogd replied:
Thanks, fixed in https://github.com/vsbogd/holeyvec/pull/1/commits/59b399d9d5be3ad359438fac83c2725229004a67

vsbogd committed:
Adding relevant keywords and categories

vsbogd committed:
Fix markdown programming language identifier

vsbogd committed:
Fix documentation

Remove documentation duplicates by including README.md into doc
generated documentation. Add comment about indices stability.

vsbogd committed:
Implement FromIterator
